p2p/src/authenticated/actors/tracker/actor.rs

```diff
@@ -499,20 +546,42 @@ impl<E: Spawner + Rng + Clock + GClock + Metrics, C: Scheme> Actor<E, C> {
                     self.store_peer_set(index, peers);
                     self.tracked_peers.set(self.peers.len() as i64);
                 }
-                Message::Reserve { peer, reservation } => {
-                    // Get latest peer set
-                    if self.allowed(&peer) {
+                Message::Reserve {
+                    public_key,
+                    reservation,
+                } => {
+                    // Block reservations for self, missing, or blocked peers
+                    if !self.valid_peer(&public_key) {
+                        debug!(?public_key, "peer not authorized to connect");
+                        let _ = reservation.send(None);
+                    } else {
                         // Because dropping the reservation will release the connection,
                         // we don't need to worry about the case that this fails.
-                        let _ = reservation.send(self.reserve(peer));
-                    } else {
-                        debug!(?peer, "peer not authorized to connect");
-                        let _ = reservation.send(None);
+                        let _ = reservation.send(self.reserve(public_key));
                     }
                 }
-                Message::Release { peer } => {
-                    self.connections.remove(&peer);
+                Message::Release { public_key } => {
+                    self.reserved.remove(&public_key);
                     self.reserved_connections.dec();
+                    self.active.remove(&public_key);
+                }
+                Message::Block { public_key } => {
+                    // Cannot block self
+                    if public_key == self.crypto.public_key() {
+                        error!(?public_key, "cannot block self");
+                        continue;
+                    }
+
+                    // Kill peer if it exists
+                    if let Some(mut peer) = self.active.remove(&public_key) {
+                        debug!(?public_key, "peer is connected, sending kill signal");
+                        peer.kill().await;
+                    }
+
+                    // Mark peer as blocked
```

Comment:
I think we should call this before `kill()`.

If not, there is a chance we try to recreate it when `await`?